polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp

```diff
@@ -788,37 +993,130 @@ void LoopInternalization::transform(T loop,
+
+  // Populate indexes needed for loading the accesses from global memory.
+  SmallVector<Value> globalIndexes(indexes.size());
+  for (int dim = indexes.size() - 1; dim >= 0; --dim) {
+    Value index = indexes[dim];
+    if (usesValue(index, IV)) {
+      Value LB = loopInfo.getLowerBound();
+      OpBuilder::InsertionGuard insertGuard(builder);
+      builder.setInsertionPointAfterValue(LB);
+      Value adjustedGlobalIV = builder.create<arith::AddIOp>(
+          localIDs[dim].getLoc(), localIDs[dim], LB);
+      IRMapping mapping;
+      mapping.map(IV, adjustedGlobalIV);
+      globalIndexes[dim] = castToIndex(
+          builder.clone(*index.getDefiningOp(), mapping)->getResult(0));
+      continue;
+    }
+    if (!loop.isDefinedOutsideOfLoop(indexes[dim]))
+      loop.moveOutOfLoop(indexes[dim].getDefiningOp());
+    globalIndexes[dim] = castToIndex(indexes[dim]);
+  }
+
+  // Load from global memory.
+  builder.setInsertionPoint(loop);
+  sycl::AccessorType accTy = getAccessorType(accessAccSub);
+  const auto idTy = cast<sycl::IDType>(
+      cast<sycl::AccessorImplDeviceType>(accTy.getBody()[0]).getBody()[0]);
+  TypedValue<MemRefType> id =
+      sycl::constructSYCLID(idTy, globalIndexes, builder, loc);
+  const Value zeroIndex = builder.create<arith::ConstantIndexOp>(loc, 0);
+  Value accSub = sycl::createSYCLAccessorSubscriptOp(accessAccSub.getAcc(), id,
+                                                     builder, loc);
+  auto load = builder.create<memref::LoadOp>(loc, accSub, zeroIndex);
+
+  // Store to local memory.
+  SmallVector<Value> reversedLocalIDs(localIDs);
+  std::reverse(reversedLocalIDs.begin(), reversedLocalIDs.end());
+  builder.create<memref::StoreOp>(loc, load, localMemory, reversedLocalIDs);
+
+  // Populate indexes use in loop with local memory.
+  SmallVector<Value> adjustedIndexes;
+  for (int dim = indexes.size() - 1; dim >= 0; --dim) {
+    Value index = indexes[dim];
+    if (usesValue(index, IV)) {
+      OpBuilder::InsertionGuard insertGuard(builder);
+      builder.setInsertionPointAfterValue(index);
+      IRMapping mapping;
+      Value adjustedIV = loopInfo.getAdjustedIV();
+      mapping.map(IV, adjustedIV);
+      adjustedIndexes.push_back(castToIndex(
+          builder.clone(*index.getDefiningOp(), mapping)->getResult(0)));
+      continue;
+    }
+    adjustedIndexes.push_back(localIDs[dim]);
+  }
+
+  // Replace original global accesses with local accesses.
+  SmallVector<Operation *> users(memref->getUsers());
+  for (Operation *user : users) {
+    OpBuilder::InsertionGuard insertGuard(builder);
+    builder.setInsertionPoint(user);
+    assert(isa<affine::AffineLoadOp>(user) && "Expecting affine load user");
+    auto load = builder.create<memref::LoadOp>(user->getLoc(), localMemory,
+                                               adjustedIndexes);
+    user->replaceAllUsesWith(load);
+    user->erase();
+  }
+
+  if (memref->use_empty())
```

Comment:
I think this should be an assert because all the loads from global memory should have been replaced with loads from shared local memory.